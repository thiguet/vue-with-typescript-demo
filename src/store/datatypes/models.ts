export enum Measures {
    UN = 'UN',
    L = 'L',
    KG = 'KG',
    G = 'G',
}

export interface Product {
    name: string;
    measure: keyof typeof Measures;
    qtd: number;
    minQtd: number;
    img?: string;
}

export interface User {
    name: string;
    email: string;
}

export interface LoginSubmit {
    name: string;
    pass: string;
}

export enum VuexAppModules {
    products = 'products',
    login = 'login',
    alert = 'alert',
}
