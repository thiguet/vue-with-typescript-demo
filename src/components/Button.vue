<template>
    <div class="wrapper-btn">
        <button :id="id" class="btn" :name="name" @click.prevent="onclick">
            <img class="icon" v-if="icon" :src="icon" :alt="name" />
            <span v-else>{{ label }} </span>
        </button>
    </div>
</template>

<script lang="ts">
import Vue from 'vue';

import { Prop, Component } from 'vue-property-decorator';

@Component
export default class Button extends Vue {
    @Prop({ required: true }) readonly id!: string;

    @Prop({ required: true }) readonly name!: string;

    @Prop({ required: true }) readonly onclick!: Function;

    @Prop() readonly label!: string;

    @Prop() readonly icon!: string;
}
</script>

<style scoped>
.wrapper-btn {
    display: flex;
    align-self: center;
    justify-content: center;
}

button {
    font-family: var(--font-family);
    font-size: var(--font-input-size);
    border: none;
    background-color: var(--second-bg);
    color: #fff;
    padding: 10px 15px;
    outline-offset: -4px;
    cursor: pointer;
}
</style>
