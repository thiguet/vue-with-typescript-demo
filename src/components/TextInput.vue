<template>
    <input
        :id="id"
        :type="type"
        :name="id"
        class="input"
        :value="value"
        @input="onInput"
    />
</template>

<script lang="ts">
import Vue from 'vue';

import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';

@Component({
    name: 'TextInput',
})
export default class TextInput extends Vue {
    @Prop({ required: true }) readonly id!: string;

    @Prop({ default: 'text' }) readonly type!: string;

    @Prop({ required: true }) readonly value!: string;

    @Prop({ required: true }) readonly setValue!: Function;

    private onInput(e: Event) {
        this.setValue((e.target as HTMLInputElement).value);
    }
}
</script>

<style scoped>
input {
    border: none;
    border-bottom: 1px solid var(--border-color);
    width: 100%;
    font-family: inherit;
    font-size: var(--font-input-size);
}
</style>
