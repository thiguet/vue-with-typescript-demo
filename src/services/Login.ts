import { LoginSubmit, User } from '../store/datatypes/models';

import VueWithTSAPI from './index';

export { VueWithTSAPI };

export const setJWT = (jwt: string) => {
    VueWithTSAPI.defaults.headers.common.Authorization = `Bearer ${jwt}`;
};

export const clearJWT = () => {
    VueWithTSAPI.defaults.headers.common.Authorization = null;
};

export async function login(user: LoginSubmit): Promise<User> {
    const response = await VueWithTSAPI.post('/users/login', {
        user,
    });
    return response.data as User;
}
