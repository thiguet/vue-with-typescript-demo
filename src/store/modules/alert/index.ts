import { Module } from 'vuex-smart-module';
import Mutations, { MutationTypes as MTypes } from './mutations';

export const MutationTypes = MTypes;
export const AlertMutations = Mutations;

export class AlertState {
  message!: string;

  display!: boolean;
}

export default new Module({
    state: AlertState,
    mutations: AlertMutations,
});
