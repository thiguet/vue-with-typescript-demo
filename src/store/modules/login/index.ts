import { Module } from 'vuex-smart-module';
import LoginMutations, { MutationTypes as MTypes } from './mutations';
import LoginActions, { ActionTypes as ATypes } from './actions';
import State from './state';

export type LoginState = State;
export const Mutations = LoginMutations;
export const MutationTypes = MTypes;
export const Actions = LoginActions;
export const ActionTypes = ATypes;

export default new Module({
    state: State,
    mutations: LoginMutations,
    actions: LoginActions,
});
