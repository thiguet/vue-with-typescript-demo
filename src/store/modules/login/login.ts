import {
  VuexModule,
  Module,
  Mutation,
  Action,
} from 'vuex-module-decorators';
import { LoginState } from './models';

@Module({
  namespaced: true,
  name: 'login',
})
export default class Login extends VuexModule implements LoginState {
  username!: string;
  password!: string;

  @Mutation
  public setUsername(newUsername: string): void {
    this.username = newUsername;
  }

  @Mutation
  public setPassword(newPass: string): void {
    this.password = newPass;
  }
}