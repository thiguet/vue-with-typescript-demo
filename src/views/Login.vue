<template>
    <div class="login">
        <form class="form">
            <InputWrapper label="Login" legend="Login">
                <TextInput
                    id="email"
                    :value="username"
                    :setValue="setUsername"
                />
            </InputWrapper>
            <InputWrapper label="Senha" legend="Password">
                <TextInput
                    id="pass"
                    type="password"
                    :value="password"
                    :setValue="setPassword"
                />
            </InputWrapper>
            <Button label="Login" :onclick="clickLogin" />
        </form>
    </div>
</template>

<script lang="ts">
import { Vue, Component } from 'vue-property-decorator';

import InputWrapper from '@/components/InputWrapper.vue';
import TextInput from '@/components/TextInput.vue';
import Button from '@/components/Button.vue';

import { namespace } from 'vuex-class';
import { LoginView } from '@/views/models.d';

import { VuexAppModules } from '@/store/datatypes/models';

const login = namespace(VuexAppModules.login);

@Component({
    name: 'Login',
    components: {
        InputWrapper,
        TextInput,
        Button,
    },
})
export default class Login extends Vue implements LoginView {
    @login.State
    private username!: string;

    @login.Mutation
    private setUsername!: (username: string) => void;

    @login.State
    private password!: string;

    @login.Mutation
    private setPassword!: (password: string) => void;

    @login.Action
    public loginAction!: () => Promise<unknown>;

    public clickLogin(): Promise<unknown> {
        return this.loginAction();
    }
}
</script>

<style scoped>
.form {
    display: flex;
    flex-flow: column;
}

.about {
    display: flex;
    flex-flow: row;
    width: 100vw;
    align-items: center;
    justify-content: center;
}

.form > fieldset {
    width: 70vw;
}

.login {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    display: flex;
    flex-wrap: wrap;
    flex-direction: column;
    align-items: center;
    overflow: hidden;
}
</style>
