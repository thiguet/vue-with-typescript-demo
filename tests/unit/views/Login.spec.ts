import { shallowMount, mount, createLocalVue, Wrapper } from '@vue/test-utils';

import Vuex, { Store, MutationTree, ActionTree } from 'vuex';

import Login from '@/views/Login.vue';

import Button from '@/components/Button.vue';
import TextInput from '@/components/TextInput.vue';

import faker from 'faker';
import { LoginState } from '@/store/modules/login';

import { LoginView } from '@/views/models.d';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('Login', () => {
    let state: LoginState;
    let mutations: MutationTree<LoginState>;
    let actions: ActionTree<MutationTree<LoginState>, LoginState>;
    let store: Store<LoginState>;

    const build = () => {
        // Now we can change state, before a running test.
        store = new Vuex.Store({
            modules: {
                login: {
                    namespaced: true,
                    state,
                    mutations,
                    actions,
                },
            },
        });

        const wrapper: Wrapper<Login> = shallowMount(Login, {
            localVue,
            store,
        });

        const mountedWrapper: Wrapper<Login> = mount(Login, {
            localVue,
            store,
        });
        return {
            wrapper,
            mountedWrapper,
            wrapperVM: (wrapper.vm as unknown) as LoginView,
            name: () => mountedWrapper.findAllComponents(TextInput).at(0),
            pass: () => mountedWrapper.findAllComponents(TextInput).at(1),
            btn: () => mountedWrapper.findComponent(Button),
        };
    };

    beforeEach(() => {
        const name = faker.internet.userName();

        state = {
            currentUser: {
                name,
                email: name,
            },
            username: name,
            password: faker.internet.password(),
        };

        mutations = {
            setUsername: jest.fn(),
            setPassword: jest.fn(),
        };

        actions = {
            loginAction: jest.fn(),
        };
    });

    it('renders component', () => {
        state = {
            currentUser: {
                name: 'testing_username',
                email: 'someemail@a.com',
            },
            username: 'testing_username',
            password: '$@3sometest',
        };
        const { wrapper } = build();
        expect(wrapper).toMatchSnapshot();
    });

    it('renders main child components', () => {
        const { name, pass, btn } = build();

        expect(name().exists()).toBe(true);
        expect(pass().exists()).toBe(true);
        expect(btn().exists()).toBe(true);
    });

    it('passes vuex props to child components', () => {
        const { name, pass, btn } = build();

        expect(name().props().value).toBe(state.username);

        expect(pass().props().value).toBe(state.password);

        expect(btn().props().onclick).toBeInstanceOf(Function);
    });

    it('submit invalid login', async () => {
        state.username = '$#@Â¨&*(guyadshj';
        const { btn } = build();
        const { loginAction } = actions;

        await btn().trigger('click');
        expect(loginAction).not.toHaveBeenCalled();
    });

    it('submit valid login', async () => {
        const { btn } = build();
        const { loginAction } = actions;
        await btn().trigger('click');
        // Used to force the waiting for the desired methods.
        await setTimeout(() => expect(loginAction).toHaveBeenCalled(), 0);
    });

    it('calling clickLogin from vm', async () => {
        const { wrapperVM } = build();
        const { loginAction } = actions;
        await wrapperVM.clickLogin();
        // Used to force the waiting for the desired methods.
        await setTimeout(() => expect(loginAction).toHaveBeenCalled(), 0);
    });

    it('must have a password input typed', async () => {
        const { pass } = build();
        expect(pass().props().type).toBe('password');
    });
});
