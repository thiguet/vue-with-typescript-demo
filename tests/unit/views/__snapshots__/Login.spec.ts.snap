// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders component 1`] = `
<div class="login">
  <form class="form">
    <inputwrapper-stub label="Login" legend="Login">
      <textinput-stub id="email" type="text" value="testing_username" setvalue="function () { [native code] }"></textinput-stub>
    </inputwrapper-stub>
    <inputwrapper-stub label="Senha" legend="Password">
      <textinput-stub id="pass" type="password" value="$@3sometest" setvalue="function () { [native code] }"></textinput-stub>
    </inputwrapper-stub>
    <button-stub label="Login" onclick="function () { [native code] }"></button-stub>
  </form>
</div>
`;
